### PR TITLE
Refactor mgmt non-LRO

### PR DESCRIPTION
**Pull request by isra-fel**
# Description
Refactor LRO writer code, now `MgmtLRO` (abbr) and `NonLRO` inherit `LRO`; `MgmtLROWriter` and `NonLROWriter` inherit `LROWriter`.

# Checklist

To ensure a quick review and merge, please ensure:
- [x] The PR has a understandable title and description explaining the _why_ and _what_.
- [ ] The PR is opened in draft if not ready for review yet.
   - If opened in draft, please allocate sufficient time (24 hours) after moving out of draft for review
- [x] The branch is recent enough to not have merge conflicts upon creation.

# Ready to Land?
- [ ] Build is completely green
   - Submissions with test failures require tracking issue and approval of a CODEOWNER
- [ ] At least one +1 review by a CODEOWNER
- [ ] All -1 reviews are confirmed resolved by the reviewer 
   - Override/Marking reviews stale must be discussed with CODEOWNERS first

### CODE REVIEW

File: src/AutoRest.CSharp/Common/Generation/Writers/LongRunningOperationWriter.cs [hunk 1]
```diff
@@ -126,7 +126,7 @@ void WriteResultFunction(bool async)
                     WriteWaitForCompletionVariants(writer, operation);
                     writer.Line();
 
-                    if (operation.ResultType != null)
+                    if (operation.ResultType != null && interfaceType != null)
```
Comment (chamons) [inline src/AutoRest.CSharp/Common/Generation/Writers/LongRunningOperationWriter.cs:129]: Why is this needed?

File: samples/Azure.Management.Storage/Generated/BlobContainersCreateOrUpdateImmutabilityPolicyOperation.cs [hunk 2]
```diff
@@ -11,30 +11,29 @@
 using Azure;
 using Azure.Core;
 using Azure.Management.Storage.Models;
-using Azure.ResourceManager.Core;
 
 namespace Azure.Management.Storage
 {
     /// <summary> Creates or updates an unlocked immutability policy. ETag in If-Match is honored if given but not required for this operation. </summary>
-    public partial class BlobContainersCreateOrUpdateImmutabilityPolicyOperation : Operation<BlobContainer>
+    public partial class BlobContainersCreateOrUpdateImmutabilityPolicyOperation : Operation<ImmutabilityPolicy>
```
Comment (isra-fel) [inline samples/Azure.Management.Storage/Generated/BlobContainersCreateOrUpdateImmutabilityPolicyOperation.cs:18]: This is an example showing not ALL PUT API should wrap the result type in [Resource], so I narrow down the condition to also detect result type == `[Resource]Data`

**Review by chamons — CHANGES_REQUESTED** [review]
I've got concerns about this PR, in particular what this line says about our class structure:

```
internal class NonLongRunningOperationWriter : LongRunningOperationWriter
```

Why do we have a non LRO using the LRO writer? If we want to share the code from LROWriter, we should be making reasonable abstractions off of it to pull out code in static helpers or other places.

File: src/AutoRest.CSharp/Common/Output/Models/LongRunningOperation.cs [hunk 3]
```diff
@@ -17,6 +16,13 @@ namespace AutoRest.CSharp.Output.Models.Requests
...
+        protected LongRunningOperation(BuildContext context) : base(context)
+        {
+            Description = string.Empty;
+            DefaultName = string.Empty;
+            DefaultAccessibility = string.Empty;
+        }
+
         public LongRunningOperation(OperationGroup operationGroup, Input.Operation operation, BuildContext context, LongRunningOperationInfo lroInfo) : base(context)
         {
             Debug.Assert(operation.IsLongRunning);
```
Comment (isra-fel) [inline src/AutoRest.CSharp/Common/Output/Models/LongRunningOperation.cs:28]: The original constructor was written based on the fact that input operation **is** LRO. However with `NonLRO` now inherits it, I found it very difficult to reuse, so I came up with an additional _protected_ constructor.

Comment (chamons) [reply src/AutoRest.CSharp/Common/Output/Models/LongRunningOperation.cs:28]: This to me suggests that this modeling is wrong, and we should rethink it.

File: src/AutoRest.CSharp/Common/Output/Models/LongRunningOperation.cs [hunk 4]
```diff
@@ -17,6 +16,13 @@ namespace AutoRest.CSharp.Output.Models.Requests
 {
     internal class LongRunningOperation : TypeProvider
     {
+        protected LongRunningOperation(BuildContext context) : base(context)
```
Comment (pakrym) [inline src/AutoRest.CSharp/Common/Output/Models/LongRunningOperation.cs:19]: This constructor leaves the object uninitialized.

**Review by chamons — APPROVED** [review]
Removing objection from landing as @m-nash is filing a tracking issue for refactoring the LRO types.
